monai/transforms/croppad/array.py

```diff
@@ -1212,3 +1213,58 @@ def __call__(self, img: NdarrayOrTensor) -> np.ndarray:
             bbox.append([i for k in zip(start_, end_) for i in k])
 
         return np.stack(bbox, axis=0)
+
+
+class PatchIter:
+    """
+    Return a patch generator with predefined properties such as `patch_size`.
+    Typically used with :py:class:`monai.data.GridPatchDataset`.
+
+    Note:
+        The `patch_size` is the size of the patch to sample from the input arrays.
+        It is assumed the arrays first dimension is the channel dimension which will be yielded in its entirety
+        so this should not be specified in `patch_size`. For example, for an input 3D array with 1 channel of size
+        (1, 20, 20, 20) a regular grid sampling of eight patches (1, 10, 10, 10) would be specified by
+        a `patch_size` of (10, 10, 10).
+
+    Args:
+        patch_size: size of patches to generate slices for, 0/None selects whole dimension
+        start_pos: starting position in the array, default is 0 for each dimension
+        mode: {``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``, ``"mean"``,
+            ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
+            One of the listed string values or a user supplied function. Defaults to ``"wrap"``.
+            See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
+        kwargs: other arguments for the `np.pad` function.
+            note that `np.pad` treats channel dimension as the first dimension.
+
+    """
+
+    backend = [TransformBackends.NUMPY]
+
+    def __init__(
+        self,
+        patch_size: Sequence[int],
+        start_pos: Sequence[int] = (),
+        mode: Union[NumpyPadMode, str] = NumpyPadMode.WRAP,
+        **kwargs,
+    ):
+        self.patch_size = (None,) + tuple(patch_size)  # expand to have the channel dim
+        self.start_pos = ensure_tuple(start_pos)
+        self.mode: NumpyPadMode = look_up_option(mode, NumpyPadMode)
+        self.kwargs = kwargs
+
+    def __call__(self, img: NdarrayOrTensor):
+        """
+        Args:
+            img: the image to generate patches from.
+
+        """
+        img_np, *_ = convert_data_type(img, np.ndarray)
+        yield from iter_patch(
```

Comment:
this is very different from the other transforms, can we keep it in the grid_dataset module?

Reply:
I think we also have other similar transforms that don't return transformed image, like `BoundingRect`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/croppad/array.py#L1175
And the `Transform` and `MapTransform` base classes don't define the return data of `__call__()` function:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/transform.py#L242
What do you think?

Thanks.

Reply:
this is a class to convert images for patch-level analysis:
```py
image_dataset = Dataset(list_of_images, transforms=...)
patch_iter = PatchIter(...)
patch_dataset = GridPatchDataset(image_dataset, patch_iter, transforms=...)
shuffled_patches = ShuffleBuffer(patch_dataset)
```
so the image/patch dataset has its own preprocessing steps, and the patchiter describes the logic of generating patches from images.

But I don't think this pipeline is very important given that we have various cropping transforms and the sliding window methods.

Reply:
OK, let me change it back to `grid` logic to be compatible with the existing logic.

Thanks.

Reply:
Hi @wyli ,

I have updated the `PatchIter` and `PatchIterd` according to your comments, could you please help review it again?

Thanks in advance.